File: bgearSort.py
from ops import s, p, r, rr
from stackops import turnedP, is_ascendingP, swap_at, switchedP, prepTop, set_ascending, set_descending, drain


def bgearSort(a, b):
    moves = 0

    if sorted(a) == a:
        return 0
    if turnedP(a):
        while turnedP(a):
            # set_ascending(a)
            r(a, "a") # rra might be more efficient in some cases
            moves += 1
        return moves

    if switchedP(a, is_ascendingP) > (-1):
            return swap_at(a, switchedP(a, is_ascendingP), "a")

    # chunks = len(a) // 2 + 1
    if len(a) < 50:
        chunks = 3
    elif len(a) <= 100:
        chunks = 7
    # if len(a) > 100:
    else:
        # chunks = len(a) // 100 * 7
        chunks = 10

    moves += prepTop(a, chunks)
    p(b, a, "b")
    moves += 1

    while a:
        if turnedP(a):
            print("Turned.")
            moves += set_ascending(a, 'a')
            print("Unturned a")
        
        if switchedP(a, is_ascendingP) > (-1):
           swap_at(a, switchedP(a, is_ascendingP), "a")
        
        # if a is in order and everything is bigger than in b
        # then there is no need to pb
        # Just order b and pa!
        if sorted(a) == a and min(a) > max(b):
            moves += drain(b, a, "a")
            print(a)
            # return moves
            break

        # Put smallest number on top by rev/rotating
        moves += prepTop(a, chunks)

        # Top of A is smaller than everything in B
        if a[0] < min(b):
            moves += set_descending(b, "b")
            p(b, a, "b")
            r(b, "b")
            moves += 2

        # Top of A is bigger than everythin in B
        elif a[0] > max(b):
            moves += set_descending(b, "b")
            p(b, a, "b")
            moves += 1

        # Top of A is not smaller nor bigger than numbers in B
        else:
            #TODO: optmize this a lot!!!
            moves += set_descending(b, "b")
            while a[0] < b[0]:
                r(b, "b")
                moves += 1
            p(b, a, "b")
            moves += 1

        # print(a)
        # print(b)


    while sorted(a) != a and b and b[0] != max(b):
        r(b, "b")
        moves += 1
    print(a)
    print(b)


    while b:
        p(a, b, "a")
        moves += 1

    print()
    print(a)
    # print(b)
    return moves

File: src/bgearSort.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   bgearSort.c                                        :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: amaria-d <amaria-d@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2022/09/27 09:08:33 by amaria-d          #+#    #+#             */
/*   Updated: 2022/09/27 10:33:11 by amaria-d         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "../header/push_swap.h"

size_t  bgearSort(t_list **a, t_list **b)
{
    size_t  moves, chunks;

    moves = 0;

    if (is_ascendingP(*a))
        return (0);
    
    if (turnedP(*a))
    {
        while (turnedP(*a))
        {
            r_p(a, 'a');
            moves++;
        }
        return (moves);
    }
    
    if (switchedP(*a, is_ascendingP) > (-1))
        return (swap_at(*a, switchedP(*a, is_ascendingP)));

    if (ft_lstlen(*a) < 50)
        chunks = 3;
    else if (ft_lstlen(*a) <= 100)
        chunks = 7;
    else
        chunks = 10;

    moves = moves + prepTop(a, chunks);
    p_p(b, a, 'b');
    moves++;

    while (*a)
    {
        if (turnedP(*a))
        {
            ft_printf("Turned__");
            moves = moves + set_ascending(a);
            ft_printf("Unturned a");
        }

        if (switchedP(*a, is_ascendingP) > (-1))
            swap_at(*a, switchedP(*a, is_ascendingP));
        
        if (is_ascendingP(*a) && (ft_lstmin(*a) > ft_lstmax(*b)))
        {
            moves = moves + drain(b, a);
            ft_lstprint(*a);
            break;
        }

        moves = moves + prepTop(a, chunks);

        // ft_lstprint(*a);
        // ft_lstprint(*b);

        if (ft_pslstget_it(*a, 0) < ft_lstmin(*b))
        {
            moves = moves + set_descending(b);
            p_p(b, a, 'a');
            r_p(b, 'b');
            moves = moves + 2;
        }

        else if (ft_pslstget_it(*a, 0) > ft_lstmax(*b))
        {
            moves = moves + set_descending(b);
            p_p(b, a, 'b');
            moves++;
        }
        else
        {
            moves = moves + set_descending(b);
            while (ft_pslstget_it(*a, 0) < ft_pslstget_it(*b, 0))
            {
                r_p(b, 'b');
                moves++;
            }
            p_p(b, a, 'b');
            moves++;
        }
    }

    while ((is_ascendingP(*a) && *b) && (ft_pslstget_it(*b, 0) != ft_lstmax(*b)))
    {
        r_p(b, 'b');
        moves++;
    }
    // ft_lstprint(*a);
    // ft_lstprint(*b);

    while (*b)
    {
        p_p(a, b, 'a');
        moves++;
    }
    
    // ft_printf("\n");
    // ft_lstprint(*a);

    return (moves);
}